File: code/java/coinChange.java
public class Solution {
    public int coinChange(int[] coins, int amount) {
        int Max = 2 * amount + 10;
        int[] dp = new int[amount+1];
        for(int i=1;i<amount+1;i++)
            dp[i] = Max;
        
        int total = 1;
        while(total <= amount){
            List<Integer> temp = new ArrayList<Integer>();
            for(int c: coins)
                if(total - c >= 0)
                    temp.add(dp[total-c] + 1);
                else
                    temp.add(Max);
            for(int j=0;j<temp.size();j++)
                if(dp[total] > temp.get(j))
                    dp[total] = temp.get(j);
            total++;
        }
        
        if(dp[amount] == Max)
            return -1;
        else
            return dp[amount];
    }

}

File: code/python/coinChange.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        Max = float('inf')
        dp = [0] + amount * [Max]
        i = 1
        
        while i <= amount:
            temp = []
            for c in coins:
                if i - c >= 0:
                    temp.append(dp[i-c] + 1) 
                else:
                    temp.append(Max)
            dp[i] = min(temp)
            i += 1
        
        if dp[amount] == Max:
            return -1
        else:
            return dp[amount]
        
